### DISCUSSION TITLE
Declarative base vs Dataclass mapping

### DISCUSSION THREAD

**Discussion by amirotin** [sqlalchemy/sqlalchemy#7929] [Usage Questions] [answered]
Hi. For a long time now I have been trying to understand the difference between using declarative_base and dataclass mapping. Except for the convenience of announcing models or personal preferences.
Are there any real advantages to using dataclass? For example in the use of slots to reduce memory usage?

**Comment by zzzeek** [comment, marked as answer]
SQLAlchemy ORM mapped classes can't use slots whether or not they are dataclasses.

basically dataclasses get you the typed constructor as well as automatic construction of things like `__repr__()`.  

However, a pure dataclass mapped with SQLAlchemy like in our current docs won't have correct typed behaviors at the class level, like "MyClass.id == 5" type of thing.

SQLAlchemy 2.0 will include a new feature that allows you to build a regular declarative mapped class that has dataclass behaviors, including the constructor; this is being made possible (well, more useful) by a new pep https://peps.python.org/pep-0681/ .  This code is in review and will be ready to go for the next major SQLAlchemy release.

short answer I certainly dont think SQLAlchemy's current dataclass integration is worth it, people really wanted it but it's very verbose and doesn't get you all the correct typing behaviors.

**Reply by amirotin** [reply]
Thanks for the answer, because I've been racking my brains trying to figure it out
